lib/filterx/func-set-timestamp.c

```diff
@@ -200,3 +200,91 @@ filterx_function_set_timestamp_new(FilterXFunctionArgs *args, GError **error)
   filterx_expr_unref(&self->super.super);
   return NULL;
 }
+
+#define FILTERX_FUNC_GET_TIMESTAMP_USAGE "Usage: get_timestamp(datetime, stamp=[\"stamp\", \"recvd\"])"
+
+typedef struct FilterXFunctionGetTimestamp_
+{
+  FilterXFunction super;
+
+  LogMessageTimeStamp timestamp_idx;
+} FilterXFunctionGetTimestamp;
+
+
+static FilterXObject *
+_get_timestamp_eval(FilterXExpr *s)
+{
+  FilterXFunctionGetTimestamp *self = (FilterXFunctionGetTimestamp *) s;
+  FilterXEvalContext *context = filterx_eval_get_context();
+  LogMessage *msg = context->msg;
+
+  return filterx_datetime_new(&msg->timestamps[self->timestamp_idx]);
+}
+
+static const gchar *
+_extract_get_timestamp_idx_str(FilterXFunctionArgs *args, GError **error)
+{
+  gboolean exists;
+  const gchar *idx_str = filterx_function_args_get_named_literal_string(args, "stamp", NULL, &exists);
+
+  if (!exists)
+    {
+      return "stamp";
+    }
+
+  return idx_str;
+}
+
+static gboolean
+_extract_get_timestamp_args(FilterXFunctionGetTimestamp *self, FilterXFunctionArgs *args, GError **error)
+{
+  gsize len = filterx_function_args_len(args);
+
+  if (len > 0)
+    {
+      g_set_error(error, FILTERX_FUNCTION_ERROR, FILTERX_FUNCTION_ERROR_CTOR_FAIL,
+                  "invalid number of arguments. " FILTERX_FUNC_GET_TIMESTAMP_USAGE);
+      return FALSE;
+    }
+
+  const gchar *idx_str = _extract_get_timestamp_idx_str(args, error);
+  if (!idx_str)
+    {
+      g_set_error(error, FILTERX_FUNCTION_ERROR, FILTERX_FUNCTION_ERROR_CTOR_FAIL,
+                  "Second argument must be string type. " FILTERX_FUNC_GET_TIMESTAMP_USAGE);
+      return FALSE;
+    }
+
+  self->timestamp_idx = log_msg_lookup_time_stamp_name(idx_str);
+
+  if ((self->timestamp_idx != LM_TS_STAMP) && (self->timestamp_idx != LM_TS_RECVD))
```

Comment:
Shouldn't we just check for the error value -1?

Reply:
set_timestamp() also only accepts these values, although I think we can allow all values.

I am not sure why that was restricted.

Reply:
I've changed this, it sems the lookup_time_stamp_name() function only supports recvd and stamp anyway.